src/Datastore/DatabaseBasedProductRepository.php

```
@@ -0,0 +1,40 @@
+<?php
+
+namespace Supermarket\Datastore;
+
+use PDO;
+
+class DatabaseBasedProductRepository implements ProductRepository
+{
+    /**
+     * @var PDO
+     */
+    private $mySqlConnection;
+
+    public function __construct(PDO $mySqlConnection)
+    {
+        $this->mySqlConnection = $mySqlConnection;
+    }
+
+    public function getAllProducts(): array
+    {
+        $fetchProducts = $this->mySqlConnection->prepare('select * from products');
+        $fetchProducts->execute();
+        $products = $fetchProducts->fetchAll(PDO::FETCH_ASSOC);
+        $productList = [];
+        foreach ($products as $product) {
+            $productList [] = new Product($product['id'], $product['name'], $product['price'], $product['type']);
+        }
+        return $productList;
+    }
+
+    public function getProductById(int $id): Product
+    {
+        $fetchProduct = $this->mySqlConnection->prepare('select * from products where id=?');
+        $fetchProduct->execute([$id]);
+        $fetchedProduct = $fetchProduct->fetch();
+        $product = new Product($fetchedProduct['id'], $fetchedProduct['name'], $fetchedProduct['price'],
```

you forgot to handle the case when the product does not exist...
please throw an exception in that case ;)
(also don't forget to catch the exception in the product details controller and return 404 in that case :) )

---

Also you could use a named constructor (factory method) for this, e.g.:
$product = Product::createFromArray($fetchedProductData);
where the createFromArray static method can do the `return new Product($fetchedProduct['id'], .... other params);`